src/libraries/System.Threading.RateLimiting/src/System/Threading/RateLimiting/ChainedPartitionedRateLimiter.cs

```diff
@@ -0,0 +1,188 @@
+            {
+                RateLimitLease lease;
+                try
+                {
+                    lease = await _limiters[i].WaitAsync(resourceID, permitCount, cancellationToken).ConfigureAwait(false);
+                }
+                catch (Exception ex)
+                {
+                    Exception? innerEx = CommonDispose(leases, i);
+                    if (innerEx is AggregateException aggregateException)
+                    {
+                        Exception[] exceptions = new Exception[aggregateException.InnerExceptions.Count + 1];
+                        aggregateException.InnerExceptions.CopyTo(exceptions, 0);
+                        exceptions[exceptions.Length - 1] = ex;
+                        throw new AggregateException(exceptions);
+                    }
+                    // REVIEW: Chose consistent exception type here, but could be convinced to just throw the original exception as is
+                    throw new AggregateException(ex);
+                }
+                if (!lease.IsAcquired)
+                {
+                    Exception? ex = CommonDispose(leases, i);
+                    return ex is not null ? throw ex : lease;
+                }
+
+                leases ??= new RateLimitLease[_limiters.Length];
+                leases[i] = lease;
+            }
+
+            return new CombinedRateLimitLease(leases!);
+        }
+
+        // REVIEW: Do we dispose the inner limiters? Or just mark the object as disposed and throw in all the methods
+        protected override void Dispose(bool disposing) => base.Dispose(disposing);
+        protected override ValueTask DisposeAsyncCore() => base.DisposeAsyncCore();
+
+        // Common dispose logic for leases when calling Acquire or WaitAsync and one of the limiters throws or can't be acquired at this time
+        private static Exception? CommonDispose(RateLimitLease[]? leases, int i)
+        {
+            List<Exception>? exceptions = null;
+            while (i > 0)
+            {
+                i--;
+                try
+                {
+                    leases![i].Dispose();
+                }
+                catch (Exception ex)
+                {
+                    exceptions ??= new List<Exception>();
+                    exceptions.Add(ex);
+                }
+            }
+
+            if (exceptions is not null)
+            {
+                return new AggregateException(exceptions);
+            }
+
+            return null;
+        }
+
+        private sealed class CombinedRateLimitLease : RateLimitLease
+        {
+            private RateLimitLease[]? _leases;
+
+            public CombinedRateLimitLease(RateLimitLease[] leases)
+            {
+                _leases = leases;
+            }
+
+            public override bool IsAcquired => true;
+
+            public override IEnumerable<string> MetadataNames => throw new NotImplementedException();
```

Comment:
Why not?

```suggestion
            public override IEnumerable<string> MetadataNames => _leases.SelectMany(l => l.MetadataNames).Distinct();
```

With less Linq of course.